File: cleanlink-android/app/src/main/java/com/example/cleanlink/LinkCleaner.kt
package com.example.cleanlink

import java.net.URL

/** Cleans social/media URLs, providing alt frontends for X/Twitter when possible. */
object LinkCleaner {
    data class CleanResult(
        val original: String,
        val twitter: String? = null,
        val vxTwitter: String? = null,
        val fxTwitter: String? = null,
        val generic: String? = null
    ) {
        /** Preferred single link for sharing/opening (twitter primary else generic). */
        val preferred: String
            get() = twitter ?: generic ?: original
    }

    private val timeRegex = Regex("^[0-9]+([smh])?$")

    fun clean(raw: String): CleanResult? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null
        val url = try { URL(extractUrlString(trimmed)) } catch (e: Exception) { return null }
        val host = url.host.lowercase().removePrefix("www.")

        // Twitter/X logic
        if (host == "x.com" || host == "twitter.com") {
            val statusPath = extractStatusPath(url.toString()) ?: return null
            val twitter = "https://x.com$statusPath"
            val vx = "https://vxtwitter.com$statusPath"
            val fx = "https://fxtwitter.com$statusPath"
            return CleanResult(original = trimmed, twitter = twitter, vxTwitter = vx, fxTwitter = fx, generic = twitter)
        }
        // Instagram: strip query
        if (host == "instagram.com") {
            val clean = "${url.protocol}://${url.host}${url.path}".removeSuffix("/")
            return CleanResult(original = trimmed, generic = clean)
        }
        // YouTube rules: preserve v id (video) and valid t; for youtu.be short links map path to v param
        if (host == "youtube.com" || host == "youtu.be") {
            var vVal: String? = null
            if (host == "youtu.be") {
                // youtu.be/<id> -> id before any further path segments
                val shortId = url.path.trim('/').split('/').firstOrNull()?.takeWhile { it.isLetterOrDigit() || it == '-' || it == '_' }
                if (!shortId.isNullOrEmpty()) vVal = shortId
            }
            val comps = url.query?.split('&') ?: emptyList()
            var tVal: String? = null
            for (p in comps) {
                val parts = p.split('=')
                if (parts.size == 2) {
                    when(parts[0]) {
                        "v" -> if (vVal == null) vVal = parts[1]
                        "t" -> if (timeRegex.matches(parts[1])) tVal = parts[1]
                    }
                }
            }
            val params = mutableListOf<String>()
            if (vVal != null) params += "v=$vVal"
            if (tVal != null) params += "t=$tVal"
            val suffix = if (params.isNotEmpty()) "?" + params.joinToString("&") else ""
            val clean = "${url.protocol}://${url.host}${url.path.takeWhile { it != '?' && it != '#' }}$suffix"
            return CleanResult(original = trimmed, generic = clean)
        }
        // Threads: strip query
        if (host == "threads.net" || host == "threads.com") {
            val clean = "${url.protocol}://${url.host}${url.path}".removeSuffix("/")
            return CleanResult(original = trimmed, generic = clean)
        }
        // Generic fallback: remove all query params
        val clean = "${url.protocol}://${url.host}${url.path}".removeSuffix("/")
        return CleanResult(original = trimmed, generic = clean)
    }

    /** Extract first URL-like token from raw share text (handles multiple lines). */
    private fun extractUrlString(input: String): String {
        val urlRegex = Regex("https?://[\n\r\t \u0000-\u007F]+", RegexOption.IGNORE_CASE)
        val match = urlRegex.find(input)
        return match?.value?.split(' ', '\n', '\r', '\t')?.firstOrNull { it.startsWith("http") } ?: input
    }

    private fun extractStatusPath(urlStr: String): String? {
        return try {
            val u = URL(urlStr)
            var host = u.host.lowercase()
            host = host.removePrefix("m.").removePrefix("mobile.")
            if (host != "twitter.com" && host != "x.com") return null
            val parts = u.path.split('/').filter { it.isNotBlank() }
            val statusIdx = parts.indexOfFirst { it.equals("status", ignoreCase = true) }
            if (statusIdx == -1 || statusIdx == parts.lastIndex) return null
            val id = parts[statusIdx + 1].takeWhile { it.isDigit() }
            if (id.isEmpty()) return null
            
            val user = parts.firstOrNull() ?: return null
            
            
            
            
            "./$user/status/$id".removePrefix("./").let { "/$user/status/$id" }
        } catch (e: Exception) { null }
    }
}

File: cleanlink-ios/CleanLinkShareExtension/LinkCleaner.swift
import Foundation

struct CleanResult {
    let original: String
    let twitter: String?
    let vxTwitter: String?
    let fxTwitter: String?
    let generic: String?
    var preferred: String { twitter ?? generic ?? original }
}

enum LinkCleaner {
    private static let timeRegex = try! NSRegularExpression(pattern: "^[0-9]+([smh])?$", options: [])

    static func clean(_ raw: String) -> CleanResult? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        guard let urlStr = extractUrlString(trimmed), let url = URL(string: urlStr) else { return nil }
        guard let host = url.host?.lowercased().replacingOccurrences(of: "www.", with: "") else { return nil }

        if host == "x.com" || host == "twitter.com" || host.hasSuffix(".twitter.com") {
            if let statusPath = extractStatusPath(url.absoluteString) {
                let twitter = "https://x.com" + statusPath
                let vx = "https://vxtwitter.com" + statusPath
                let fx = "https://fxtwitter.com" + statusPath
                return CleanResult(original: trimmed, twitter: twitter, vxTwitter: vx, fxTwitter: fx, generic: twitter)
            } else { return nil }
        }
        if host == "instagram.com" {
            let clean = (url.scheme ?? "https") + "://" + host + url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            return CleanResult(original: trimmed, twitter: nil, vxTwitter: nil, fxTwitter: nil, generic: clean)
        }
        if host == "youtube.com" || host == "youtu.be" {
            var vVal: String? = nil
            if host == "youtu.be" {
                let shortId = url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/")).components(separatedBy: "/").first?.filter { $0.isLetter || $0.isNumber || $0 == "-" || $0 == "_" }
                if let sid = shortId, !sid.isEmpty { vVal = sid }
            }
            var tVal: String? = nil
            if let query = url.query {
                for part in query.split(separator: "&") {
                    let kv = part.split(separator: "=")
                    if kv.count == 2 {
                        let key = String(kv[0])
                        let val = String(kv[1])
                        switch key {
                        case "v": if vVal == nil { vVal = val }
                        case "t": if timeRegex.matches(in: val, options: [], range: NSRange(location: 0, length: val.count)).count == 1 { tVal = val }
                        default: break
                        }
                    }
                }
            }
            var params: [String] = []
            if let vVal = vVal { params.append("v=\(vVal)") }
            if let tVal = tVal { params.append("t=\(tVal)") }
            let suffix = params.isEmpty ? "" : "?" + params.joined(separator: "&")
            let pathNoQuery = url.path.split(separator: "?").first.map(String.init) ?? url.path
            let clean = (url.scheme ?? "https") + "://" + (url.host ?? host) + pathNoQuery + suffix
            return CleanResult(original: trimmed, twitter: nil, vxTwitter: nil, fxTwitter: nil, generic: clean)
        }
        if host == "threads.net" || host == "threads.com" {
            let clean = (url.scheme ?? "https") + "://" + host + url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            return CleanResult(original: trimmed, twitter: nil, vxTwitter: nil, fxTwitter: nil, generic: clean)
        }
        let clean = (url.scheme ?? "https") + "://" + host + url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return CleanResult(original: trimmed, twitter: nil, vxTwitter: nil, fxTwitter: nil, generic: clean)
    }

    private static func extractUrlString(_ input: String) -> String? {
        let pattern = "https?://[A-Za-z0-9./?=&_%:-]+"
        if let range = input.range(of: pattern, options: .regularExpression) {
            return String(input[range])
        }
        return input.starts(with: "http") ? input : nil
    }

    private static func extractStatusPath(_ urlStr: String) -> String? {
        guard let u = URL(string: urlStr), var host = u.host?.lowercased() else { return nil }
        host = host.replacingOccurrences(of: "m.", with: "").replacingOccurrences(of: "mobile.", with: "")
        guard host == "twitter.com" || host == "x.com" else { return nil }
        let parts = u.path.split(separator: "/").map(String.init).filter { !$0.isEmpty }
        guard let statusIdx = parts.firstIndex(where: { $0.lowercased() == "status" }), statusIdx < parts.count - 1 else { return nil }
        let id = parts[statusIdx + 1].prefix { $0.isNumber }
        guard !id.isEmpty else { return nil }
        guard let user = parts.first else { return nil }
        return "/\(user)/status/\(id)"
    }
}
